File: sdk/typescript/src/behaviors.ts
import type { CallOptions } from "@connectrpc/connect";

import type { SoulFireBot } from "./client.js";
import {
  BlockFace,
  Hand,
} from "./generated/soulfire/bot_live_pb.js";
import type { BlockPosition } from "./generated/soulfire/common_pb.js";
import {
  BotTaskStatus,
  type BotTaskEvent,
} from "./generated/soulfire/task_pb.js";

export interface BehaviorContext {
  signal?: AbortSignal;
}

export interface BotBehavior<T = void> {
  run(bot: SoulFireBot, context?: BehaviorContext): Promise<T>;
}

export type BehaviorResult<Behavior> =
  Behavior extends BotBehavior<infer Result> ? Result : never;

export type BehaviorResults<
  Behaviors extends readonly BotBehavior<unknown>[],
> = {
  readonly [Index in keyof Behaviors]: BehaviorResult<Behaviors[Index]>;
};

export class SoulFireBehaviorError extends Error {
  public constructor(
    public readonly behavior: string,
    message: string,
    options?: ErrorOptions,
  ) {
    super(message, options);
    this.name = "SoulFireBehaviorError";
  }
}

export class SoulFireBehaviorTimeoutError extends SoulFireBehaviorError {
  public constructor(public readonly durationMs: number) {
    super("timeout", `Behavior exceeded ${durationMs} ms`);
    this.name = "SoulFireBehaviorTimeoutError";
  }
}

export function defineBehavior<T>(
  run: BotBehavior<T>["run"],
): BotBehavior<T> {
  return { run };
}

export async function runBehaviors(
  bot: SoulFireBot,
  behaviors: readonly BotBehavior[],
  context?: BehaviorContext,
): Promise<void> {
  for (const behavior of behaviors) {
    context?.signal?.throwIfAborted();
    await behavior.run(bot, context);
  }
}

export function sequence<
  const Behaviors extends readonly BotBehavior<unknown>[],
>(
  ...behaviors: Behaviors
): BotBehavior<BehaviorResults<Behaviors>> {
  return defineBehavior(async (bot, context) => {
    const results: unknown[] = [];
    for (const behavior of behaviors) {
      context?.signal?.throwIfAborted();
      results.push(await behavior.run(bot, context));
    }
    return results as BehaviorResults<Behaviors>;
  });
}

export interface ParallelOptions {
  readonly concurrency?: number;
}

export function parallel<
  const Behaviors extends readonly BotBehavior<unknown>[],
>(
  behaviors: Behaviors,
  options: ParallelOptions = {},
): BotBehavior<BehaviorResults<Behaviors>> {
  const concurrency = options.concurrency === undefined
    ? behaviors.length || 1
    : positiveInteger(options.concurrency, "concurrency");
  return defineBehavior(async (bot, context) => {
    const link = linkedAbortController(context?.signal);
    try {
      const results = await mapConcurrent(
        behaviors,
        concurrency,
        (behavior) => behavior.run(bot, { signal: link.controller.signal }),
      );
      return results as BehaviorResults<Behaviors>;
    } catch (error) {
      link.controller.abort(error);
      throw error;
    } finally {
      link.dispose();
    }
  });
}

export function race<
  const Behaviors extends readonly [
    BotBehavior<unknown>,
    ...BotBehavior<unknown>[],
  ],
>(
  ...behaviors: Behaviors
): BotBehavior<BehaviorResults<Behaviors>[number]> {
  return defineBehavior(async (bot, context) => {
    const link = linkedAbortController(context?.signal);
    try {
      return await Promise.any(
        behaviors.map((behavior) =>
          behavior.run(bot, { signal: link.controller.signal })
        ),
      ) as BehaviorResults<Behaviors>[number];
    } finally {
      link.controller.abort("Behavior race completed");
      link.dispose();
    }
  });
}

export interface RepeatOptions {
  readonly times: number;
}

export function repeat<T>(
  behavior: BotBehavior<T>,
  options: RepeatOptions,
): BotBehavior<readonly T[]> {
  const times = positiveInteger(options.times, "times");
  return defineBehavior(async (bot, context) => {
    const results: T[] = [];
    for (let iteration = 0; iteration < times; iteration += 1) {
      context?.signal?.throwIfAborted();
      results.push(await behavior.run(bot, context));
    }
    return results;
  });
}

export interface RetryOptions {
  readonly attempts?: number;
  readonly delayMs?: number;
  readonly backoff?: number;
  readonly maximumDelayMs?: number;
}

export function retry<T>(
  behavior: BotBehavior<T>,
  options: RetryOptions = {},
): BotBehavior<T> {
  const attempts = positiveInteger(options.attempts ?? 3, "attempts");
  const initialDelay = nonNegativeFinite(options.delayMs ?? 0, "delayMs");
  const backoff = positiveFinite(options.backoff ?? 1, "backoff");
  const maximumDelay = nonNegativeFinite(
    options.maximumDelayMs ?? Number.MAX_SAFE_INTEGER,
    "maximumDelayMs",
  );
  return defineBehavior(async (bot, context) => {
    let delayMs = initialDelay;
    for (let attempt = 1; attempt <= attempts; attempt += 1) {
      context?.signal?.throwIfAborted();
      try {
        return await behavior.run(bot, context);
      } catch (error) {
        if (attempt === attempts || context?.signal?.aborted === true) {
          throw error;
        }
        await abortableDelay(delayMs, context?.signal);
        delayMs = Math.min(delayMs * backoff, maximumDelay);
      }
    }
    throw new AssertionError("Retry loop exhausted unexpectedly");
  });
}

export function timeout<T>(
  behavior: BotBehavior<T>,
  durationMs: number,
): BotBehavior<T> {
  const duration = positiveFinite(durationMs, "durationMs");
  return defineBehavior(async (bot, context) => {
    const link = linkedAbortController(context?.signal);
    let timer: ReturnType<typeof setTimeout> | undefined;
    const timeoutError = new SoulFireBehaviorTimeoutError(duration);
    const expired = new Promise<never>((_, reject) => {
      timer = setTimeout(() => {
        link.controller.abort(timeoutError);
        reject(timeoutError);
      }, duration);
    });
    try {
      return await Promise.race([
        behavior.run(bot, { signal: link.controller.signal }),
        expired,
      ]);
    } finally {
      if (timer !== undefined) {
        clearTimeout(timer);
      }
      link.dispose();
    }
  });
}

export interface UntilOptions {
  readonly maximumIterations?: number;
}

export type BehaviorPredicate<T> = (
  result: T,
) => boolean | Promise<boolean>;

export function until<T>(
  behavior: BotBehavior<T>,
  predicate: BehaviorPredicate<T>,
  options: UntilOptions = {},
): BotBehavior<T> {
  const maximumIterations = options.maximumIterations === undefined
    ? Number.MAX_SAFE_INTEGER
    : positiveInteger(options.maximumIterations, "maximumIterations");
  return defineBehavior(async (bot, context) => {
    for (
      let iteration = 1;
      iteration <= maximumIterations;
      iteration += 1
    ) {
      context?.signal?.throwIfAborted();
      const result = await behavior.run(bot, context);
      if (await predicate(result)) {
        return result;
      }
    }
    throw new SoulFireBehaviorError(
      "until",
      `Predicate remained false after ${maximumIterations} iterations`,
    );
  });
}

export type BotPredicate = (
  bot: SoulFireBot,
) => boolean | Promise<boolean>;

export function conditional<WhenTrue, WhenFalse = void>(
  predicate: BotPredicate,
  whenTrue: BotBehavior<WhenTrue>,
  whenFalse?: BotBehavior<WhenFalse>,
): BotBehavior<WhenTrue | WhenFalse> {
  return defineBehavior(async (bot, context) =>
    await predicate(bot)
      ? whenTrue.run(bot, context)
      : whenFalse === undefined
        ? undefined as WhenFalse
        : whenFalse.run(bot, context)
  );
}

export function fallback<T>(
  primary: BotBehavior<T>,
  ...alternatives: readonly BotBehavior<T>[]
): BotBehavior<T> {
  return defineBehavior(async (bot, context) => {
    const failures: unknown[] = [];
    for (const behavior of [primary, ...alternatives]) {
      context?.signal?.throwIfAborted();
      try {
        return await behavior.run(bot, context);
      } catch (error) {
        failures.push(error);
      }
    }
    throw new AggregateError(failures, "Every fallback behavior failed");
  });
}

export function cleanup<T>(
  behavior: BotBehavior<T>,
  finalizer: BotBehavior<unknown>,
): BotBehavior<T> {
  return defineBehavior(async (bot, context) => {
    let failure: unknown;
    try {
      return await behavior.run(bot, context);
    } catch (error) {
      failure = error;
      throw error;
    } finally {
      try {
        await finalizer.run(bot, context);
      } catch (finalizerError) {
        if (failure !== undefined) {
          throw new AggregateError(
            [failure, finalizerError],
            "Behavior and cleanup both failed",
          );
        }
        throw finalizerError;
      }
    }
  });
}

export function scopedLease<T>(
  behavior: BotBehavior<T>,
  ttlSeconds = 30,
): BotBehavior<T> {
  const ttl = positiveInteger(ttlSeconds, "ttlSeconds");
  return defineBehavior(async (bot, context) => {
    const lease = await bot.acquireControl(ttl, callOptions(context));
    try {
      return await behavior.run(bot, context);
    } finally {
      await lease.release(callOptions(context));
    }
  });
}

export interface CollectBlocksOptions {
  blockIds: readonly string[];
  tags?: readonly string[];
  count?: number;
  searchRadius?: number;
  allowPlacing?: boolean;
}

export function collectBlocks(
  options: CollectBlocksOptions,
): BotBehavior<number> {
  return defineBehavior(async (bot, context) => {
    const call = callOptions(context);
    const task = await bot.tasks.collectBlocks(options.blockIds, {
      tags: options.tags ?? [],
      count: options.count ?? 1,
      searchRadius: options.searchRadius ?? 32,
      path: {
        allowMining: true,
        allowPlacing: options.allowPlacing ?? false,
      },
      ...(call === undefined ? {} : { call }),
    });
    return (await task.result(
      call === undefined ? {} : { call },
    )).blocksBroken;
  });
}

export function followEntity(
  entityId: number,
  radius = 3,
): BotBehavior<void> {
  return defineBehavior(async (bot, context) => {
    const call = callOptions(context);
    await completeTask(bot.tasks.runFollowEntity(
      entityId,
      radius,
      {
        path: {
          allowMining: false,
          allowPlacing: false,
        },
        ...(call === undefined
          ? {}
          : { call }),
      },
    ));
  });
}

export interface AttackNearestOptions {
  entityTypes: readonly string[];
  radius?: number;
  attackRange?: number;
  sprinting?: boolean;
  maximumAttacks?: number;
}

export function attackNearest(
  options: AttackNearestOptions,
): BotBehavior<boolean> {
  return defineBehavior(async (bot, context) => {
    const response = await bot.listNearbyEntities(
      {
        entityTypes: [...options.entityTypes],
        includePlayers: false,
        radius: options.radius ?? 32,
      },
      callOptions(context),
    );
    const target = response.entities[0];
    if (target === undefined) {
      return false;
    }
    const call = callOptions(context);
    await completeTask(bot.tasks.runAttackEntity(
      target.entityId,
      {
        attackRange: options.attackRange ?? 3,
        sprinting: options.sprinting ?? false,
        maximumAttacks: options.maximumAttacks ?? 0,
        path: {
          allowMining: false,
          allowPlacing: false,
        },
        ...(call === undefined ? {} : { call }),
      },
    ));
    return true;
  });
}

export interface AutoEatOptions {
  foodItemIds: readonly string[];
  foodLevel?: number;
  checkIntervalTicks?: number;
  maximumMeals?: number;
  completeWhenNoFood?: boolean;
  restoreSelectedSlot?: boolean;
}

export function autoEat(options: AutoEatOptions): BotBehavior<void> {
  return defineBehavior(async (bot, context) => {
    const call = callOptions(context);
    await completeTask(
      bot.tasks.runAutoEat(options.foodItemIds, {
        foodLevel: options.foodLevel ?? 14,
        checkIntervalTicks: options.checkIntervalTicks ?? 20,
        maximumMeals: options.maximumMeals ?? 0,
        completeWhenNoFood: options.completeWhenNoFood ?? false,
        restoreSelectedSlot: options.restoreSelectedSlot ?? true,
        ...(call === undefined ? {} : { call }),
      }),
      "autoEat",
    );
  });
}

export interface AutoRespawnOptions {
  respawnDelayTicks?: number;
  maximumRespawns?: number;
}

export function autoRespawn(
  options: AutoRespawnOptions = {},
): BotBehavior<void> {
  return defineBehavior(async (bot, context) => {
    const call = callOptions(context);
    await completeTask(
      bot.tasks.runAutoRespawn({
        respawnDelayTicks: options.respawnDelayTicks ?? 0,
        maximumRespawns: options.maximumRespawns ?? 0,
        ...(call === undefined ? {} : { call }),
      }),
      "autoRespawn",
    );
  });
}

export interface AutoTotemOptions {
  checkIntervalTicks?: number;
  maximumEquips?: number;
  completeWhenNoTotem?: boolean;
  replaceOccupiedOffhand?: boolean;
}

export function autoTotem(
  options: AutoTotemOptions = {},
): BotBehavior<void> {
  return defineBehavior(async (bot, context) => {
    const call = callOptions(context);
    await completeTask(
      bot.tasks.runAutoTotem({
        checkIntervalTicks: options.checkIntervalTicks ?? 20,
        maximumEquips: options.maximumEquips ?? 0,
        completeWhenNoTotem: options.completeWhenNoTotem ?? false,
        replaceOccupiedOffhand: options.replaceOccupiedOffhand ?? false,
        ...(call === undefined ? {} : { call }),
      }),
      "autoTotem",
    );
  });
}

export interface AutoArmorOptions {
  checkIntervalTicks?: number;
  maximumEquips?: number;
  completeWhenNoUpgrade?: boolean;
}

export function autoArmor(
  options: AutoArmorOptions = {},
): BotBehavior<void> {
  return defineBehavior(async (bot, context) => {
    const call = callOptions(context);
    await completeTask(
      bot.tasks.runAutoArmor({
        checkIntervalTicks: options.checkIntervalTicks ?? 20,
        maximumEquips: options.maximumEquips ?? 0,
        completeWhenNoUpgrade: options.completeWhenNoUpgrade ?? false,
        ...(call === undefined ? {} : { call }),
      }),
      "autoArmor",
    );
  });
}

export interface BuildPlacement {
  against: BlockPosition;
  face: BlockFace;
  hotbarSlot?: number;
}

export function build(
  placements: readonly BuildPlacement[],
): BotBehavior<number> {
  return defineBehavior(async (bot, context) => {
    let placed = 0;
    for (const placement of placements) {
      context?.signal?.throwIfAborted();
      if (placement.hotbarSlot !== undefined) {
        await bot.selectHotbar(placement.hotbarSlot, callOptions(context));
      }
      await bot.placeBlock(
        {
          against: placement.against,
          face: placement.face,
          hand: Hand.MAIN,
        },
        callOptions(context),
      );
      placed += 1;
    }
    return placed;
  });
}

async function completeTask(
  events: AsyncIterable<BotTaskEvent>,
  behavior = "task",
): Promise<void> {
  let last: BotTaskEvent | undefined;
  for await (const event of events) {
    last = event;
  }
  const task = last?.task;
  if (task?.status === BotTaskStatus.COMPLETED) {
    return;
  }
  throw new Error(
    task?.failure?.message
      ?? `${behavior} ended without successful completion`,
  );
}

function callOptions(context?: BehaviorContext): CallOptions | undefined {
  return context?.signal === undefined
    ? undefined
    : { signal: context.signal };
}

async function mapConcurrent<Input, Output>(
  values: readonly Input[],
  concurrency: number,
  operation: (value: Input) => Promise<Output>,
): Promise<Output[]> {
  const results = new Array<Output>(values.length);
  let next = 0;
  const worker = async () => {
    while (next < values.length) {
      const index = next;
      next += 1;
      results[index] = await operation(values[index] as Input);
    }
  };
  await Promise.all(
    Array.from(
      { length: Math.min(concurrency, values.length) },
      worker,
    ),
  );
  return results;
}

function linkedAbortController(parent?: AbortSignal): {
  controller: AbortController;
  dispose: () => void;
} {
  const controller = new AbortController();
  if (parent === undefined) {
    return { controller, dispose: () => undefined };
  }
  if (parent.aborted) {
    controller.abort(parent.reason);
    return { controller, dispose: () => undefined };
  }
  const abort = () => controller.abort(parent.reason);
  parent.addEventListener("abort", abort, { once: true });
  return {
    controller,
    dispose: () => parent.removeEventListener("abort", abort),
  };
}

function abortableDelay(
  durationMs: number,
  signal?: AbortSignal,
): Promise<void> {
  if (durationMs === 0) {
    signal?.throwIfAborted();
    return Promise.resolve();
  }
  return new Promise((resolve, reject) => {
    signal?.throwIfAborted();
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", abort);
      resolve();
    }, durationMs);
    const abort = () => {
      clearTimeout(timer);
      reject(signal?.reason);
    };
    signal?.addEventListener("abort", abort, { once: true });
  });
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new RangeError(`${name} must be a positive safe integer`);
  }
  return value;
}

function nonNegativeFinite(value: number, name: string): number {
  if (!Number.isFinite(value) || value < 0) {
    throw new RangeError(`${name} must be a finite non-negative number`);
  }
  return value;
}

function positiveFinite(value: number, name: string): number {
  if (!Number.isFinite(value) || value <= 0) {
    throw new RangeError(`${name} must be a finite positive number`);
  }
  return value;
}

class AssertionError extends Error {
  public override readonly name = "AssertionError";
}

File: sdk/python/src/soulfire/behaviors.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable, Iterable, Sequence
from dataclasses import dataclass
from typing import Protocol, TypeVar

from .bot import AsyncSoulFireBot
from .bot_live_pb2 import (
    HAND_MAIN,
    BlockFace,
    PathfindOptions,
)
from .common_pb2 import BlockPosition

ResultT = TypeVar("ResultT", covariant=True)


class BotBehavior(Protocol[ResultT]):
    async def run(self, bot: AsyncSoulFireBot) -> ResultT: ...


@dataclass(frozen=True, slots=True)
class FunctionBehavior[ResultT]:
    function: Callable[[AsyncSoulFireBot], Awaitable[ResultT]]

    async def run(self, bot: AsyncSoulFireBot) -> ResultT:
        return await self.function(bot)


def define_behavior[ResultT](
    function: Callable[[AsyncSoulFireBot], Awaitable[ResultT]],
) -> FunctionBehavior[ResultT]:
    return FunctionBehavior(function)


class SoulFireBehaviorError(RuntimeError):
    pass


class SoulFireBehaviorTimeoutError(SoulFireBehaviorError):
    def __init__(self, duration: float) -> None:
        self.duration = duration
        super().__init__(f"Behavior exceeded {duration:g} seconds")


async def run_behaviors(
    bot: AsyncSoulFireBot,
    behaviors: Iterable[BotBehavior[object]],
) -> None:
    for behavior in behaviors:
        await behavior.run(bot)


@dataclass(frozen=True, slots=True)
class SequenceBehavior:
    behaviors: tuple[BotBehavior[object], ...]

    async def run(self, bot: AsyncSoulFireBot) -> tuple[object, ...]:
        return tuple([await behavior.run(bot) for behavior in self.behaviors])


def sequence(*behaviors: BotBehavior[object]) -> SequenceBehavior:
    return SequenceBehavior(behaviors)


@dataclass(frozen=True, slots=True)
class ParallelBehavior:
    behaviors: tuple[BotBehavior[object], ...]

    async def run(self, bot: AsyncSoulFireBot) -> tuple[object, ...]:
        results: list[object | None] = [None] * len(self.behaviors)

        async def run_one(index: int, behavior: BotBehavior[object]) -> None:
            results[index] = await behavior.run(bot)

        async with asyncio.TaskGroup() as tasks:
            for index, behavior in enumerate(self.behaviors):
                tasks.create_task(run_one(index, behavior))
        return tuple(results)


def parallel(*behaviors: BotBehavior[object]) -> ParallelBehavior:
    return ParallelBehavior(behaviors)


@dataclass(frozen=True, slots=True)
class RaceBehavior:
    behaviors: tuple[BotBehavior[object], ...]

    async def run(self, bot: AsyncSoulFireBot) -> object:
        if not self.behaviors:
            raise ValueError("race requires at least one behavior")
        tasks = {asyncio.create_task(behavior.run(bot)) for behavior in self.behaviors}
        failures: list[Exception] = []
        try:
            pending = tasks
            while pending:
                completed, pending = await asyncio.wait(
                    pending,
                    return_when=asyncio.FIRST_COMPLETED,
                )
                for task in completed:
                    try:
                        result = task.result()
                    except Exception as error:
                        failures.append(error)
                    else:
                        for remaining in pending:
                            remaining.cancel()
                        return result
            raise ExceptionGroup("Every raced behavior failed", failures)
        finally:
            for task in tasks:
                if not task.done():
                    task.cancel()
            await asyncio.gather(*tasks, return_exceptions=True)


def race(
    first: BotBehavior[object],
    *others: BotBehavior[object],
) -> RaceBehavior:
    return RaceBehavior((first, *others))


@dataclass(frozen=True, slots=True)
class RepeatBehavior[ResultT]:
    behavior: BotBehavior[ResultT]
    times: int

    async def run(self, bot: AsyncSoulFireBot) -> tuple[ResultT, ...]:
        times = _positive_integer(self.times, "times")
        return tuple([await self.behavior.run(bot) for _ in range(times)])


def repeat[ResultT](
    behavior: BotBehavior[ResultT],
    *,
    times: int,
) -> RepeatBehavior[ResultT]:
    _positive_integer(times, "times")
    return RepeatBehavior(behavior, times)


@dataclass(frozen=True, slots=True)
class RetryBehavior[ResultT]:
    behavior: BotBehavior[ResultT]
    attempts: int = 3
    delay: float = 0
    backoff: float = 1
    maximum_delay: float | None = None

    async def run(self, bot: AsyncSoulFireBot) -> ResultT:
        attempts = _positive_integer(self.attempts, "attempts")
        delay = _non_negative_finite(self.delay, "delay")
        backoff = _positive_finite(self.backoff, "backoff")
        maximum_delay = (
            float("inf")
            if self.maximum_delay is None
            else _non_negative_finite(self.maximum_delay, "maximum_delay")
        )
        for attempt in range(1, attempts + 1):
            try:
                return await self.behavior.run(bot)
            except Exception:
                if attempt == attempts:
                    raise
                await asyncio.sleep(delay)
                delay = min(delay * backoff, maximum_delay)
        raise AssertionError("Retry loop exhausted unexpectedly")


def retry[ResultT](
    behavior: BotBehavior[ResultT],
    *,
    attempts: int = 3,
    delay: float = 0,
    backoff: float = 1,
    maximum_delay: float | None = None,
) -> RetryBehavior[ResultT]:
    return RetryBehavior(
        behavior,
        attempts=attempts,
        delay=delay,
        backoff=backoff,
        maximum_delay=maximum_delay,
    )


@dataclass(frozen=True, slots=True)
class TimeoutBehavior[ResultT]:
    behavior: BotBehavior[ResultT]
    seconds: float

    async def run(self, bot: AsyncSoulFireBot) -> ResultT:
        duration = _positive_finite(self.seconds, "seconds")
        try:
            async with asyncio.timeout(duration):
                return await self.behavior.run(bot)
        except TimeoutError as error:
            raise SoulFireBehaviorTimeoutError(duration) from error


def timeout[ResultT](
    behavior: BotBehavior[ResultT],
    seconds: float,
) -> TimeoutBehavior[ResultT]:
    _positive_finite(seconds, "seconds")
    return TimeoutBehavior(behavior, seconds)


@dataclass(frozen=True, slots=True)
class UntilBehavior[ResultT]:
    behavior: BotBehavior[ResultT]
    predicate: Callable[[ResultT], bool | Awaitable[bool]]
    maximum_iterations: int | None = None

    async def run(self, bot: AsyncSoulFireBot) -> ResultT:
        maximum = (
            None
            if self.maximum_iterations is None
            else _positive_integer(self.maximum_iterations, "maximum_iterations")
        )
        iteration = 0
        while maximum is None or iteration < maximum:
            result = await self.behavior.run(bot)
            iteration += 1
            if await _resolve_bool(self.predicate(result)):
                return result
        raise SoulFireBehaviorError(f"Predicate remained false after {maximum} iterations")


def until[ResultT](
    behavior: BotBehavior[ResultT],
    predicate: Callable[[ResultT], bool | Awaitable[bool]],
    *,
    maximum_iterations: int | None = None,
) -> UntilBehavior[ResultT]:
    return UntilBehavior(behavior, predicate, maximum_iterations)


@dataclass(frozen=True, slots=True)
class ConditionalBehavior[ResultT]:
    predicate: Callable[[AsyncSoulFireBot], bool | Awaitable[bool]]
    when_true: BotBehavior[ResultT]
    when_false: BotBehavior[ResultT] | None = None

    async def run(self, bot: AsyncSoulFireBot) -> ResultT | None:
        if await _resolve_bool(self.predicate(bot)):
            return await self.when_true.run(bot)
        if self.when_false is None:
            return None
        return await self.when_false.run(bot)


def conditional[ResultT](
    predicate: Callable[[AsyncSoulFireBot], bool | Awaitable[bool]],
    when_true: BotBehavior[ResultT],
    when_false: BotBehavior[ResultT] | None = None,
) -> ConditionalBehavior[ResultT]:
    return ConditionalBehavior(predicate, when_true, when_false)


@dataclass(frozen=True, slots=True)
class FallbackBehavior[ResultT]:
    behaviors: tuple[BotBehavior[ResultT], ...]

    async def run(self, bot: AsyncSoulFireBot) -> ResultT:
        failures: list[Exception] = []
        for behavior in self.behaviors:
            try:
                return await behavior.run(bot)
            except Exception as error:
                failures.append(error)
        raise ExceptionGroup("Every fallback behavior failed", failures)


def fallback[ResultT](
    primary: BotBehavior[ResultT],
    *alternatives: BotBehavior[ResultT],
) -> FallbackBehavior[ResultT]:
    return FallbackBehavior((primary, *alternatives))


@dataclass(frozen=True, slots=True)
class CleanupBehavior[ResultT]:
    behavior: BotBehavior[ResultT]
    finalizer: BotBehavior[object]

    async def run(self, bot: AsyncSoulFireBot) -> ResultT:
        try:
            result = await self.behavior.run(bot)
        except Exception as behavior_error:
            try:
                await self.finalizer.run(bot)
            except Exception as finalizer_error:
                raise ExceptionGroup(
                    "Behavior and cleanup both failed",
                    [behavior_error, finalizer_error],
                ) from None
            raise
        await self.finalizer.run(bot)
        return result


def cleanup[ResultT](
    behavior: BotBehavior[ResultT],
    finalizer: BotBehavior[object],
) -> CleanupBehavior[ResultT]:
    return CleanupBehavior(behavior, finalizer)


@dataclass(frozen=True, slots=True)
class ScopedLeaseBehavior[ResultT]:
    behavior: BotBehavior[ResultT]
    ttl_seconds: int = 30

    async def run(self, bot: AsyncSoulFireBot) -> ResultT:
        ttl = _positive_integer(self.ttl_seconds, "ttl_seconds")
        async with await bot.acquire_control(ttl_seconds=ttl):
            return await self.behavior.run(bot)


def scoped_lease[ResultT](
    behavior: BotBehavior[ResultT],
    *,
    ttl_seconds: int = 30,
) -> ScopedLeaseBehavior[ResultT]:
    return ScopedLeaseBehavior(behavior, ttl_seconds)


@dataclass(frozen=True, slots=True)
class CollectBlocks:
    block_ids: Sequence[str]
    tags: Sequence[str] = ()
    count: int = 1
    search_radius: int = 32
    allow_placing: bool = False

    async def run(self, bot: AsyncSoulFireBot) -> int:
        task = await bot.tasks.collect_blocks(
            self.block_ids,
            tags=self.tags,
            count=self.count,
            search_radius=self.search_radius,
            options=PathfindOptions(
                allow_mining=True,
                allow_placing=self.allow_placing,
            ),
        )
        return (await task.result()).blocks_broken


@dataclass(frozen=True, slots=True)
class FollowEntity:
    entity_id: int
    radius: float = 3

    async def run(self, bot: AsyncSoulFireBot) -> None:
        async for _ in bot.tasks.run_follow_entity(
            self.entity_id,
            distance=self.radius,
            options=PathfindOptions(allow_mining=False, allow_placing=False),
        ):
            pass


@dataclass(frozen=True, slots=True)
class AttackNearest:
    entity_types: Sequence[str]
    radius: float = 32
    attack_range: float = 3
    sprinting: bool = False
    maximum_attacks: int = 0

    async def run(self, bot: AsyncSoulFireBot) -> bool:
        response = await bot.list_nearby_entities(
            self.radius,
            entity_types=self.entity_types,
            include_players=False,
        )
        if not response.entities:
            return False
        target = response.entities[0]
        async for _ in bot.tasks.run_attack_entity(
            target.entity_id,
            attack_range=self.attack_range,
            sprinting=self.sprinting,
            maximum_attacks=self.maximum_attacks,
            options=PathfindOptions(allow_mining=False, allow_placing=False),
        ):
            pass
        return True


@dataclass(frozen=True, slots=True)
class AutoEat:
    food_item_ids: Sequence[str]
    food_level: int = 14
    check_interval_ticks: int = 20
    maximum_meals: int = 0
    complete_when_no_food: bool = False
    restore_selected_slot: bool = True

    async def run(self, bot: AsyncSoulFireBot) -> None:
        async for _ in bot.tasks.run_auto_eat(
            self.food_item_ids,
            food_level=self.food_level,
            check_interval_ticks=self.check_interval_ticks,
            maximum_meals=self.maximum_meals,
            complete_when_no_food=self.complete_when_no_food,
            restore_selected_slot=self.restore_selected_slot,
        ):
            pass


@dataclass(frozen=True, slots=True)
class AutoRespawn:
    respawn_delay_ticks: int = 0
    maximum_respawns: int = 0

    async def run(self, bot: AsyncSoulFireBot) -> None:
        async for _ in bot.tasks.run_auto_respawn(
            respawn_delay_ticks=self.respawn_delay_ticks,
            maximum_respawns=self.maximum_respawns,
        ):
            pass


@dataclass(frozen=True, slots=True)
class AutoTotem:
    check_interval_ticks: int = 20
    maximum_equips: int = 0
    complete_when_no_totem: bool = False
    replace_occupied_offhand: bool = False

    async def run(self, bot: AsyncSoulFireBot) -> None:
        async for _ in bot.tasks.run_auto_totem(
            check_interval_ticks=self.check_interval_ticks,
            maximum_equips=self.maximum_equips,
            complete_when_no_totem=self.complete_when_no_totem,
            replace_occupied_offhand=self.replace_occupied_offhand,
        ):
            pass


@dataclass(frozen=True, slots=True)
class AutoArmor:
    check_interval_ticks: int = 20
    maximum_equips: int = 0
    complete_when_no_upgrade: bool = False

    async def run(self, bot: AsyncSoulFireBot) -> None:
        async for _ in bot.tasks.run_auto_armor(
            check_interval_ticks=self.check_interval_ticks,
            maximum_equips=self.maximum_equips,
            complete_when_no_upgrade=self.complete_when_no_upgrade,
        ):
            pass


@dataclass(frozen=True, slots=True)
class BuildPlacement:
    against: BlockPosition
    face: BlockFace
    hotbar_slot: int | None = None


@dataclass(frozen=True, slots=True)
class Build:
    placements: Sequence[BuildPlacement]

    async def run(self, bot: AsyncSoulFireBot) -> int:
        placed = 0
        for placement in self.placements:
            if placement.hotbar_slot is not None:
                await bot.select_hotbar(placement.hotbar_slot)
            await bot.place_block(
                placement.against,
                placement.face,
                HAND_MAIN,
            )
            placed += 1
        return placed


async def _resolve_bool(value: bool | Awaitable[bool]) -> bool:
    if isinstance(value, bool):
        return value
    return await value


def _positive_integer(value: int, name: str) -> int:
    if value <= 0:
        raise ValueError(f"{name} must be positive")
    return value


def _non_negative_finite(value: float, name: str) -> float:
    if value < 0 or value == float("inf") or value != value:
        raise ValueError(f"{name} must be a finite non-negative number")
    return value


def _positive_finite(value: float, name: str) -> float:
    if value <= 0 or value == float("inf") or value != value:
        raise ValueError(f"{name} must be a finite positive number")
    return value
